README.md
<!-- <img src="assets/image/logo.png" alt="Practical Linux logo" title="Practical Linux" width="250" align="right" /> -->
<img src="https://s2.loli.net/2022/04/16/2q84cIWknSEfMAK.png" alt="Practical Linux logo" title="Practical Linux" width="250" align="right" />

<!-- 标签生成网站 https://img.shields.io  -->

<!-- ![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg) -->
![GitHub license](https://img.shields.io/badge/license-CC%204.0-blue)
![Awesome](https://awesome.re/badge.svg)
![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/practical-linux-skills)
![GitHub commit activity](https://img.shields.io/github/commit-activity/y/baiyongan/practical-linux-skills?color=orange)
[![bya-hexo](https://img.shields.io/badge/blog-bya%40hexo-lightgrey)](https://baiyongan.github.io)
[![bya-tiddlywiki](https://img.shields.io/badge/wiki-bya%40tiddly-9cf)](http://106.14.160.45:8081)

<!-- [![Awesome](https://awesome.re/badge.svg)](https://awesome.re) ![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg) -->
<!-- [![PRs welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://github.com/amzn/style-dictionary/blob/master/CONTRIBUTING.md#submitting-pull-requests)
<br/>
[![Build Status](https://img.shields.io/travis/amzn/style-dictionary.svg?style=flat-square)](https://travis-ci.org/amzn/style-dictionary)
[![downloads](https://img.shields.io/npm/dm/style-dictionary.svg?style=flat-square)](https://www.npmjs.com/package/style-dictionary) -->

# Practical Linux Skills

> **从实践中来，到实践中去**

> *from practice, to practice*

## 写作初衷

自从接触 Linux 系统以来，从小白一路跌跌撞撞，零零碎碎地积攒经验，如今已经可以算是 Linux 系统的熟练操作者。但是，当想继续提升能力，从而实现 level up 时，却发现翻来覆去也就会那么几个常见套路。若仅凭自身的技术储备，来突破瓶颈，显得有些力不从心。故而，想做一个关于 Linux 实用技能的文档整理网站，广泛借鉴高手们的经验总结，并一点点吸收内化为自己的技能。输出日志，以作备忘的同时，也顺道提升下在业内的知名度，是为此。

自评没足够实力及精力连续输出高质量，成体系的内容 (非不愿也，实不能也)。不过，也还是有眼光及途径，来发现并整理一些有价值的最佳实践，系列专题的，希望借此方式，也能够帮助到同样需要进阶的 Linux 使用者。

## 预期目标

- 理解掌握一些必要的底层概念，背景知识，做到知其然，知其所以然；
- 辩证地吸收高手的经验，操作习惯，官方的编码规范等，提升命令行操作的流畅性，及脚本编码的规范性与严谨性；
- 总结出一些针对具体业务的最佳实践，解决方案等，尤其是脚本编程，性能调优等；
- 搜集体验各种酷炫惊艳，同时又实用高效的命令行工具等；
- 系统梳理几个主流且必备的基础技能专题，如编辑器 vim，文本处理三剑客（sed/awk/grep），版本控制工具 git/svn，容器 docker/k8s 等；
- 整理出若干套不错的教程，记忆卡片等。

## 内容简介

**总体偏向于 Linux 系统运维管理，脚本开发，及性能诊断，调优方向。** 以搜集整理网络资源，总结学习笔记为主，个人经验总结，内容点评为辅。暂定为以下几个部分:

<!-- > **1. Minimum Necessary Knowledge**

- 最少必要知识，一些应该形成肌肉记忆，条件反射的操作或概念。 -->

> **1. Command & Shell**

- 一些需要重点掌握的命令，以及其组合
- 一些提高编程效率的代码片段总结等

> **2. Best Practice**

- 一些操作上的好习惯，必要规范等
- 一些经过实践检验，个人真正在践行的案例等

> **3. Awesome Tricks & Tools**

- 一些真正使用起来，便相见恨晚的高效技法、工具等

> **4. Essential Tools Series**

- vim/VS Code
- sed/awk/grep/regular expression
- git/svn
- docker/k8s
- sql
- ansible
- ...

> **5. Linux Operating System**

- 了解操作系统内核工作原理
- 理解程序运行机制

> **6. Performance Tuning**

- 性能调优，尤见功力

## 写作工具

- mkdocs: 文档生成工具
- github pages: 托管平台
- VS Code: markdown 编写工具
- PicGo + SM.MS: 图床工具
- asciinema: 命令行录屏软件
- Anki: 高效记忆工具

## 写作方式

- 中英文混杂
- 笔记翻译，知识点注疏为主，原创输出为辅，其中，夹杂一些简要的点评和扩展
- 主要以实用为主，翻译及总结尽力而为，如需准确理解，建议浏览相关的原始资料进行学习掌握

## 关于作者

| 网站                                                         | 内容简介                                                     | 最近动态                                                     |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| [**BYA's Blog**](https://baiyongan.github.io)                | 个人博客，以读书书单及书摘为主，暂时没备案，也没优化访问体验。 | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/baiyongan.github.io?logo=github&style=for-the-badge) |
| [**BYA's Wiki**](http://106.14.160.45:8081)                  | 个人 wiki，以实用的、技术导向的课程总结为主。                | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/everything-notes?color=blue&logo=github&style=for-the-badge) |
| [**LSF Handbook**](http://baiyongan.github.io/lsf-handbook)  | LSF 官方文档的部分翻译，及个人经验总结与资源分享。           | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/lsf-handbook?logo=github&style=for-the-badge) |
| [**Practical Linux Skills**](http://106.14.160.45)           | 关于 Linux 的实用技能总结，框架已建好，择时填坑。            | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/practical-linux-skills?color=blue&logo=github&style=for-the-badge) |
| [**Advanced Python Techniques**](http://106.14.160.45:8080)  | 关于 Python 的进阶及高级知识点总结，日拱一卒，重点投入。     | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/advanced-python-techniques?color=gold&logo=github&style=for-the-badge) |
<!-- | [**IC Design Platform Best Practices**](http://106.14.160.45:8090) | 关于 IC 行业设计平台设计、构建、优化的最佳实践。             | ![GitHub last commit](https://img.shields.io/github/last-commit/baiyongan/ic-design-platform-best-practices?color=grey&logo=github&style=for-the-badge) | -->


## 版权信息

本站点采用[知识共享署名-非商业性使用 4.0 国际许可协议](https://github.com/baiyongan/practical_linux_skills/blob/main/docs/LICENSE)进行许可。

docs/CHANGELOG.md
# 更新日志

<!---------------------------------------------------------->
### 0.3.2 (2023-04-28)
#### Features
- 添加 asciinema-player plugins
  
<!---------------------------------------------------------->
### 0.3.2 (2023-04-04)
#### Articles
- 更新 performance tuning article55-58
  
<!---------------------------------------------------------->
### 0.3.1 (2023-03-15)
#### Articles
- 添加 iptables in part07
  
<!---------------------------------------------------------->
### 0.3.1 (2023-03-15)
#### Features
- 调整 part 结构
- 添加 contents & overview
#### Articles
- 添加 permission control 章节 in part07

<!---------------------------------------------------------->
### 0.3.0 (2023-03-11)
#### Features
- mkdocs 替换 docsify

<!---------------------------------------------------------->
### 0.2.36 (2021-08-04)
#### Articles
- 更新 linux operating system article 06

<!---------------------------------------------------------->
### 0.2.35 (2021-07-27)
#### Articles
- 更新 linux operating system article 06

<!---------------------------------------------------------->
### 0.2.34 (2021-07-26)
#### Articles
- 更新 bash pitfalls article 22

<!---------------------------------------------------------->
### 0.2.33 (2021-07-24)
#### Articles
- 更新 bash pitfalls article 19 & 21

<!---------------------------------------------------------->
### 0.2.32 (2021-07-22)
#### Articles
- 更新 linux operating system article 05

<!---------------------------------------------------------->
### 0.2.31 (2021-07-20)
#### Features
- 补全 bash pitfalls article till article 62
#### Articles
- 更新 bash pitfalls article 20

<!---------------------------------------------------------->
### 0.2.30 (2021-07-18)
#### Features
- 补全 bash pitfalls article till article 30
#### Articles
- 更新 bash pitfalls article 16 & 17 & 18

<!---------------------------------------------------------->
### 0.2.29 (2021-07-10)
#### Articles
- 更新 linux operating system article 05

<!---------------------------------------------------------->
### 0.2.28 (2021-07-09)
#### Articles
- 更新 linux operating system article 04

<!---------------------------------------------------------->
### 0.2.27 (2021-07-07)
#### Articles
- 更新 linux operating system article 03

<!---------------------------------------------------------->
### 0.2.26 (2021-07-06)
#### Features
- 互换 linux operating system & performance tuning
#### Articles
- 更新 linux operating system article 02

<!---------------------------------------------------------->
### 0.2.25 (2021-07-05)
#### Features
- 更新 linux operating system 目录
- 更新 vim 图片

<!---------------------------------------------------------->
### 0.2.24 (2021-07-04)
#### Articles
- 更新 bash pitfalls article 13 & 14 & 15

<!---------------------------------------------------------->
### 0.2.23 (2021-07-02)
#### Articles
- 更新 linux operating system article 00 & 01 

<!---------------------------------------------------------->
### 0.2.22 (2021-06-28)
#### Articles
- 更新 bash pitfalls article 10 & 11

<!---------------------------------------------------------->
### 0.2.21 (2021-06-27)
#### Features
- 更新 linux operating system 目录
#### Articles
- 更新 bash pitfalls article 08 & 09
  
<!---------------------------------------------------------->
### 0.2.21 (2021-06-24)
#### Features
- 添加 linux operating system 目录

<!---------------------------------------------------------->
### 0.2.20 (2021-06-23)
#### Features
- 拆分 filename pathname article 为十部分

#### Articles
- 更新 filename pathname article section 2 & 校对文字
- 删减 shell 十三问文章内容

<!---------------------------------------------------------->
### 0.2.19 (2021-06-18)
#### Features
- 新增 Quick Reference 部分
- 新增 wiki & blog 徽章

<!---------------------------------------------------------->
### 0.2.18 (2021-06-17)
#### Features
- 新增 Linux Operating System 部分
- 移除 Minimum Necessary Knowledge 部分

<!---------------------------------------------------------->
### 0.2.17 (2021-06-16)
#### Features
- 关闭 Valine
#### Articles
- 更新 bash pitfalls article 07 & 08
- 更新 bash pitfalls index

<!---------------------------------------------------------->
### 0.2.16 (2021-06-15)
#### Features
- 调整 index.html
- 变更 引用本地 js & css
#### Articles
- 更新 filename pathname article section 10
  
<!---------------------------------------------------------->
### 0.2.15 (2021-06-10)
#### Features
- 新增 徽章
- 调整 assets
#### Articles
- 更新 filename pathname article section 2 

<!---------------------------------------------------------->
### 0.2.14 (2021-06-09)
#### Features
- 新增 文字统计插件
- 新增 Valine 评论插件（与 TOC 有冲突）
- 新增 Tab 标签页 （有冲突）
- 新增 FEATURE.md
#### Articles
- 更新 filename pathname article

<!---------------------------------------------------------->
### 0.2.11 (2021-06-07)
#### Articles
- 更新 filename pathname article

<!---------------------------------------------------------->
### 0.2.10 (2021-06-01)
#### Articles
- 新增 IFS article
- 更新 REFERENCE

<!---------------------------------------------------------->
### 0.2.9 (2021-05-29)
#### Features
- 部署到 gitee
- 添加 best_practice 分支

#### Articles
- 更新 filename pathname article

<!---------------------------------------------------------->
### 0.2.7 (2021-05-25)
#### Articles
- 新增 lsof 命令

<!---------------------------------------------------------->
### 0.2.6 (2021-05-24)
#### Articles
- 新增 nfs4 acl aticle 
- 更新 bash pitfalls index 中的 $ 为 \$

<!---------------------------------------------------------->
### 0.2.5 (2021-05-15)
#### Articles
- 更新 performance tuning article 15 & 16
- 更新 linux eval 命令用法

<!---------------------------------------------------------->
### 0.2.4 (2021-04-14)
#### Articles
- 更新 performance tuning article 15 & 16

<!---------------------------------------------------------->
### 0.2.3 (2021-04-08)
#### Features
- 更改 To do list

#### Articles
- 更新 performance tuning article 12

<!---------------------------------------------------------->
### 0.2.2 (2021-03-30)
#### Features
- 完善 内嵌 pdf 插件
- 添加 LaTex 插件

#### Articles
- 更新 performance tuning article 12


assets/js/docsify/plugins/toc/toc.js
var defaultOptions = {
  headings: 'h1, h2',
  scope: '.markdown-section',

  // To make work
  title: 'Contents',
  listType: 'ul',  
}

// Element builders
var tocHeading = function(Title) {
  return document.createElement('h2').appendChild(
    document.createTextNode(Title)
  )
}

var aTag = function(src) {
  var a = document.createElement('a');
  var content = src.firstChild.innerHTML;

  // Use this to clip text w/ HTML in it.
  // https://github.com/arendjr/text-clipper
  a.innerHTML = content;
  a.href = src.firstChild.href;
  a.onclick = tocClick

  // In order to remove this gotta fix the styles.
  a.setAttribute('class', 'anchor');

  return a
};

var tocClick = function(e) {
  var divs = document.querySelectorAll('.page_toc .active');

  // Remove the previous classes
  [].forEach.call(divs, function(div) {
    div.setAttribute('class', 'anchor')
  });

  // Make sure this is attached to the parent not itself
  e.target.parentNode.setAttribute('class', 'active')
};

var createList = function(wrapper, count) {
  while (count--) {
    wrapper = wrapper.appendChild(
      document.createElement('ul')
    );

    if (count) {
      wrapper = wrapper.appendChild(
        document.createElement('li')
      );
    }
  }

  return wrapper;
};

//------------------------------------------------------------------------

var getHeaders = function(selector) {
  var headings2 = document.querySelectorAll(selector);
  var ret = [];

  [].forEach.call(headings2, function(heading) {
    ret = ret.concat(heading);
  });

  return ret;
};

var getLevel = function(header) {
  var decs = header.match(/\d/g);

  return decs ? Math.min.apply(null, decs) : 1;
};

var jumpBack = function(currentWrapper, offset) {
  while (offset--) {
    currentWrapper = currentWrapper.parentElement;
  }

  return currentWrapper;
};

var buildTOC = function(options) {
  var ret = document.createElement('ul');
  var wrapper = ret;
  var lastLi = null;
  var selector = options.scope + ' ' + options.headings
  var headers = getHeaders(selector)

  headers.reduce(function(prev, curr, index) {
    var currentLevel = getLevel(curr.tagName);
    var offset = currentLevel - prev;

    wrapper = (offset > 0)
      ? createList(lastLi, offset)
      : jumpBack(wrapper, -offset * 2)

    wrapper = wrapper || ret;

    var li = document.createElement('li');

    wrapper.appendChild(li).appendChild(aTag(curr));

    lastLi = li;

    return currentLevel;
  }, getLevel(options.headings));

  return ret;
};

// Docsify plugin functions
function plugin(hook, vm) {
  var userOptions = vm.config.toc;

  hook.mounted(function () {
    var content = window.Docsify.dom.find(".content");
    if (content) {
      var nav = window.Docsify.dom.create("aside", "");
      window.Docsify.dom.toggleClass(nav, "add", "nav");
      window.Docsify.dom.before(content, nav);
    }
  });

  hook.doneEach(function () {
    var nav = document.querySelectorAll('.nav')[0]
    var t = Array.from(document.querySelectorAll('.nav'))

    if (!nav) {
      return;
    }

  	const toc = buildTOC(userOptions);

    // Just unset it for now.
    if (!toc.innerHTML) {
      nav.innerHTML = null
      return;
    }

    // Fix me in the future
		var title = document.createElement('p');
		title.innerHTML = userOptions.title;
		title.setAttribute('class', 'title');

		var container = document.createElement('div');
		container.setAttribute('class', 'page_toc');
		
		container.appendChild(title);
		container.appendChild(toc);

    // Existing TOC
    var tocChild = document.querySelectorAll('.nav .page_toc');

    if (tocChild.length > 0) {
      tocChild[0].parentNode.removeChild(tocChild[0]);
    }

    nav.appendChild(container);
  });
}

// Docsify plugin options
window.$docsify['toc'] = Object.assign(defaultOptions, window.$docsify['toc']);
window.$docsify.plugins = [].concat(plugin, window.$docsify.plugins);


assets/js/docsify/plugins/mermaid/docsify-mermaid.js
// modules are defined as an array
// [ module function, map of requires ]
//
// map of requires is short require name -> numeric require
//
// anything defined in a previous bundle is accessed via the
// orig method which is the require for previous bundles
parcelRequire = (function (modules, cache, entry, globalName) {
  // Save the require from previous bundle to this closure if any
  var previousRequire = typeof parcelRequire === 'function' && parcelRequire;
  var nodeRequire = typeof require === 'function' && require;

  function newRequire(name, jumped) {
    if (!cache[name]) {
      if (!modules[name]) {
        // if we cannot find the module within our internal map or
        // cache jump to the current global require ie. the last bundle
        // that was added to the page.
        var currentRequire = typeof parcelRequire === 'function' && parcelRequire;
        if (!jumped && currentRequire) {
          return currentRequire(name, true);
        }

        // If there are other bundles on this page the require from the
        // previous one is saved to 'previousRequire'. Repeat this as
        // many times as there are bundles until the module is found or
        // we exhaust the require chain.
        if (previousRequire) {
          return previousRequire(name, true);
        }

        // Try the node require function if it exists.
        if (nodeRequire && typeof name === 'string') {
          return nodeRequire(name);
        }

        var err = new Error('Cannot find module \'' + name + '\'');
        err.code = 'MODULE_NOT_FOUND';
        throw err;
      }

      localRequire.resolve = resolve;
      localRequire.cache = {};

      var module = cache[name] = new newRequire.Module(name);

      modules[name][0].call(module.exports, localRequire, module, module.exports, this);
    }

    return cache[name].exports;

    function localRequire(x){
      return newRequire(localRequire.resolve(x));
    }

    function resolve(x){
      return modules[name][1][x] || x;
    }
  }

  function Module(moduleName) {
    this.id = moduleName;
    this.bundle = newRequire;
    this.exports = {};
  }

  newRequire.isParcelRequire = true;
  newRequire.Module = Module;
  newRequire.modules = modules;
  newRequire.cache = cache;
  newRequire.parent = previousRequire;
  newRequire.register = function (id, exports) {
    modules[id] = [function (require, module) {
      module.exports = exports;
    }, {}];
  };

  var error;
  for (var i = 0; i < entry.length; i++) {
    try {
      newRequire(entry[i]);
    } catch (e) {
      // Save first error but execute all entries
      if (!error) {
        error = e;
      }
    }
  }

  if (entry.length) {
    // Expose entry point to Node, AMD or browser globals
    // Based on https://github.com/ForbesLindesay/umd/blob/master/template.js
    var mainExports = newRequire(entry[entry.length - 1]);

    // CommonJS
    if (typeof exports === "object" && typeof module !== "undefined") {
      module.exports = mainExports;

    // RequireJS
    } else if (typeof define === "function" && define.amd) {
     define(function () {
       return mainExports;
     });

    // <script>
    } else if (globalName) {
      this[globalName] = mainExports;
    }
  }

  // Override the current require with this new one
  parcelRequire = newRequire;

  if (error) {
    // throw error from earlier, _after updating parcelRequire_
    throw error;
  }

  return newRequire;
})({"GOVZ":[function(require,module,exports) {
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var plugin = function plugin(hook, vm) {
  hook.afterEach(function (html, next) {
    // We load the HTML inside a DOM node to allow for manipulation
    var htmlElement = document.createElement('div');
    htmlElement.innerHTML = html;
    htmlElement.querySelectorAll('pre[data-lang=mermaid]').forEach(function (element) {
      // Create a <div class="mermaid"> to replace the <pre> 
      var replacement = document.createElement('div');
      replacement.textContent = element.textContent;
      replacement.classList.add('mermaid'); // Replace

      element.parentNode.replaceChild(replacement, element);
    });
    next(htmlElement.innerHTML);
  });
  hook.doneEach(function () {
    mermaid.init({}, '.mermaid');
  });
};

var _default = plugin;
exports.default = _default;
},{}],"Focm":[function(require,module,exports) {
"use strict";

var _plugin = _interopRequireDefault(require("./plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

if (!window.$docsify) {
  window.$docsify = {};
}

window.$docsify.plugins = (window.$docsify.plugins || []).concat(_plugin.default);
},{"./plugin":"GOVZ"}]},{},["Focm"], null)
//# sourceMappingURL=/docsify-mermaid.js.map

assets/js/docsify/plugins/pdf-embed/docsify-pdf-embed.js
const DEBUG = 0;
const PDF_VIEWER_WIDTH = '100%';
const PDF_VIEWER_HEIGHT = '50rem';
const PDF_MARGIN_TOP = '2rem';
const PDF_MARGIN_BOTTOM = '5rem';

! function() {
	var verbose_log = function(log){
		if (DEBUG) {
			console.log(log);
		}
	}

	var mobile_check = function() {
  		var check = false;
  		(function(a){if(/(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a)||/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(a.substr(0,4))) check = true;})(navigator.userAgent||navigator.vendor||window.opera);
  		return check;
	};

	var url_check = function(url_str) {
		var expression = /^(http|https):\/\/[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
		return url_str.match(expression);
	}

	var build_absolute_path = function(path_str) {
		return window.location.protocol + "//" + window.location.hostname + "/" + path_str;
	}

	var renderer_func = function(code, lang, base=null) { 
		var pdf_renderer = function(code, lang, verify) {
			function unique_id_generator(){
				function rand_gen(){
					return Math.floor((Math.random()+1) * 65536).toString(16).substring(1);
				}
				return rand_gen() + rand_gen() + '-' + rand_gen() + '-' + rand_gen() + '-' + rand_gen() + '-' + rand_gen() + rand_gen() + rand_gen();
			}
			if(lang && !lang.localeCompare('pdf', 'en', {sensitivity: 'base'})){
				if(verify){
					return true;
				}else{
					var divId = "markdown_code_pdf_container_" + unique_id_generator().toString();
					var container_list = new Array();
					if(localStorage.getItem('pdf_container_list')){
						container_list = JSON.parse(localStorage.getItem('pdf_container_list'));	
					}
					container_list.push({"pdf_location": code, "div_id": divId});
					localStorage.setItem('pdf_container_list', JSON.stringify(container_list));
					return (
						'<div style="margin-top:'+ PDF_MARGIN_TOP +'; margin-bottom:'+ PDF_MARGIN_BOTTOM +';" id="'+ divId +'">'
							+ '<a href="'+ code + '"> Link </a> to ' + code +
						'</div>'
					);
				} 
			}
			return false;
		}
		if(pdf_renderer(code, lang, true)){
		   return pdf_renderer(code, lang, false);
		}
		/* SECTION START: Put other custom code rendering functions here
			i.e. If the language of the code block is LaTex, 
			put the code below to replace original code block with the text: 
			'Using LaTex is much better than handwriting!' inside a div container.

			if (lang == "latex") {
				return ('<div class="container">Using LaTex is much better than handwriting!</div>');
			}
			
		SECTION END */
		return (base ? base : this.origin.code.apply(this, arguments));
	}

	var pdf_renderer = {code: renderer_func};
	var doc_md = window.$docsify.markdown = (window.$docsify.markdown || {});
	var doc_md_rend = doc_md.renderer = (window.$docsify.markdown.renderer || {});
	doc_md_rend.code = (doc_md_rend.code ? doc_md_rend.code : renderer_func);

	var mobile_view = mobile_check();

	// Allowing Docsify to execute the script to embed PDF
	window.$docsify.executeScript = true;

	// Linking Docsify to the PDF plugin
	window.$docsify.plugins = [
		function(hook, vm) {
			hook.init(function() {
				verbose_log('PDF Code Block Loader Plugin Initialized');
				verbose_log('This is a '.concat(mobile_view ? 'Mobile' : 'Non-mobile').concat(' Viewer.'));
			});
			hook.afterEach(function(html, next) {
				verbose_log('PDF Code Block Loader Plugin After Hook Initialized');
				container_list = JSON.parse(localStorage.getItem('pdf_container_list'));
				verbose_log(container_list);
				if(container_list){
					// Handling mobile pdf rendering
					if(mobile_view){
						html += '<script>';
						container_list.forEach(function(container){
							var absolute_pdf_location = url_check(container['pdf_location']) ? container['pdf_location'] : build_absolute_path(container['pdf_location']);
							var view_location = "https://mozilla.github.io/pdf.js/web/viewer.html?file=" + absolute_pdf_location;
							var iframe_id = container['div_id'].concat('-iframe');
							html += '\
									document.getElementById("'+ container['div_id'] +'").innerHTML = \
									\'<iframe \
										style="overflow: auto; width: '+ PDF_VIEWER_WIDTH +'; height: '+ PDF_VIEWER_HEIGHT +';"\
										id="'+ iframe_id +'"\
								    	src="'+ view_location +'"\
								    	style="border: none;" \
									\/>\'\
								;';
						});
						html += '<\/script>';
					}else{
						// Handling desktop pdf rendering
						html += '<script>';
						container_list.forEach(function(container){
							html += '\
							var options = {\
								height: "'+ PDF_VIEWER_HEIGHT +'",\
								width: "'+ PDF_VIEWER_WIDTH +'"\
							};\
							PDFObject.embed("'+ container['pdf_location'] +'", "#'+ container['div_id'] +'", options);';
						});
						html += '<\/script>';
					}
				}
				localStorage.removeItem('pdf_container_list');
				next(html);
			});
		}
	].concat(window.$docsify.plugins || [])
}();
